packages/svelte/src/structured-object.svelte.ts

```diff
@@ -0,0 +1,226 @@
+import {
+  generateId,
+  isAbortError,
+  safeValidateTypes,
+  type FetchFunction,
+} from '@ai-sdk/provider-utils';
+import {
+  asSchema,
+  isDeepEqualData,
+  parsePartialJson,
+  type DeepPartial,
+  type Schema,
+} from '@ai-sdk/ui-utils';
+import { type z } from 'zod';
+import {
+  getStructuredObjectContext,
+  hasStructuredObjectContext,
+  KeyedStructuredObjectStore,
+  type StructuredObjectStore,
+} from './structured-object-context.svelte.js';
+
+export type Experimental_StructuredObjectOptions<RESULT> = {
+  /**
+   * The API endpoint. It should stream JSON that matches the schema as chunked text.
+   */
+  api: string;
+
+  /**
+   * A Zod schema that defines the shape of the complete object.
+   */
+  schema: z.Schema<RESULT, z.ZodTypeDef, unknown> | Schema<RESULT>;
+
+  /**
+   * An unique identifier. If not provided, a random one will be
+   * generated. When provided, the `useObject` hook with the same `id` will
+   * have shared states across components.
+   */
+  id?: string;
+
+  /**
+   * An optional value for the initial object.
+   */
+  initialValue?: DeepPartial<RESULT>;
+
+  /**
+   * Custom fetch implementation. You can use it as a middleware to intercept requests,
+   * or to provide a custom fetch implementation for e.g. testing.
+   */
+  fetch?: FetchFunction;
+
+  /**
+   * Callback that is called when the stream has finished.
+   */
+  onFinish?: (event: {
+    /**
+     * The generated object (typed according to the schema).
+     * Can be undefined if the final object does not match the schema.
+     */
+    object: RESULT | undefined;
+
+    /**
+     * Optional error object. This is e.g. a TypeValidationError when the final object does not match the schema.
+     */
+    error: Error | undefined;
+  }) => Promise<void> | void;
+
+  /**
+   * Callback function to be called when an error is encountered.
+   */
+  onError?: (error: Error) => void;
+
+  /**
+   * Additional HTTP headers to be included in the request.
+   */
+  headers?: Record<string, string> | Headers;
+};
+
+export class StructuredObject<RESULT, INPUT = unknown> {
+  #options: Experimental_StructuredObjectOptions<RESULT> =
+    {} as Experimental_StructuredObjectOptions<RESULT>;
+  readonly #id = $derived(this.#options.id ?? generateId());
+  readonly #keyedStore = $state<KeyedStructuredObjectStore>()!;
```

Comment:
This also shouldn't need `$state` unless I'm missing something 😃
```suggestion
  readonly #keyedStore: KeyedStructuredObjectStore = null!
```